### ISSUE TITLE
Why is `N22_unrecoverable_isolation` considered fatal?

### ISSUE THREAD

**Issue by metalicjames** [eBay/NuRaft#223]
Under excessive load, or due to a network partition, a node could be isolated from some peers and not others (particularly the leader, in this case). If the partition resolves, the follower could reconnect and recover. Yet, this error is considered fatal and calls `state_mgr::system_exit`.

What is the rationale for this? Shouldn't the follower attempt to reconnect until success rather than give up and exit?

Here's the relevant line: https://github.com/eBay/NuRaft/blob/8715f27ac3fb3cd2fd6df946b7628123171e8340/src/raft_server.cxx#L859

**Comment by metalicjames** [comment]
Could there be a mechanism to apply replication back pressure in this case? The problem arises when there are enough nodes with sufficient bandwidth to the leader to reach commit quorum, but some other followers are unable to keep up with the growth of the log. It would be better to prevent further `append_entries` requests from queuing on the leader until all followers have been able to catch up.

**Comment by greensky00** [comment]
@metalicjames 
I'm not sure I get your concern correctly. Let me rephrase it like this: *when there are lagging followers that are not in the quorum, and cannot keep up with the incoming write speed, we want to throttle the leader to lower the acceptance rate of `append_entries` calls*, am I understanding it right?

In my opinion, throttling needs to be done outside Raft, by the caller of NuRaft APIs. And also that's how we do it in eBay. If we want to implement backpressure in the library, there are two ways: 1) returning a transient error like `TRY_LATER`, or 2) simply blocking the user threads. But neither way seems appropriate as the behavior of library API call.

**Comment by metalicjames** [comment]
Your rephrasing is correct.

If it's best done outside NuRaft, how should one detect this condition? How does eBay do it? At present, it eventually leads to a `system_exit` on the lagging follower, depending on how you set `pre_vote_rejection_limit_`.

**Comment by greensky00** [comment]
There is an API to get peer info:
https://github.com/eBay/NuRaft/blob/877a60a34d7cf3ed2fe273c05f8ff9e6fb887172/include/libnuraft/raft_server.hxx#L561-L566

You can detect such a log gap by checking `max - min > threshold` or various other ways. 

By the way, I wonder how such lagging followers end up with `N22` status, which is not a desired behavior. Even though there is a big log gap between the leader and a follower, the leader is still sending `append_entries` message not to make the follower initiate vote. Is it because a single `append_entries` message to the lagging follower gets bigger (due to the big log gap) and is not delivered on time?
